### PR TITLE
[CASSANDRA-21102] Fix off-by-one bug in exponential backoff for repair retry config

### PR DESCRIPTION
**Pull request by nivykani**
Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>

```

The [Cassandra Jira](https://issues.apache.org/jira/browse/CASSANDRA-21102)

### CODE REVIEW

File: src/java/org/apache/cassandra/service/TimeoutStrategy.java [hunk 1]
```diff
@@ -97,7 +97,7 @@ interface LatencyModifierFactory
         default LatencyModifier identity() { return (l, a) -> l; }
         default LatencyModifier multiply(double constant) { return (l, a) -> saturatedCast(l * constant); }
         default LatencyModifier multiplyByAttempts(double multiply) { return (l, a) -> saturatedCast(l * multiply * a); }
-        default LatencyModifier multiplyByAttemptsExp(double base) { return (l, a) -> saturatedCast(l * pow(base, a)); }
+        default LatencyModifier multiplyByAttemptsExp(double base) { return (l, a) -> saturatedCast(l * pow(base, max(0, a-1))); }
```
Comment (jyothsnakonisa) [inline src/java/org/apache/cassandra/service/TimeoutStrategy.java]:
```suggestion
        default LatencyModifier multiplyByAttemptsExp(double base) { return (l, a) -> saturatedCast(l * pow(base, max(0, a - 1))); }
```
Please fix formatting

File: test/unit/org/apache/cassandra/service/TimeoutStrategyTest.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+        // Attempt 3: baseLatency * 2^(3-1) = baseLatency * 4
+        Assertions.assertThat(w.getMicros(3))
+                .isEqualTo(expectedBaseLatencyMicros * 4);
+
+        // Edge case to check for 0 attempts: max(0, -1) = 0
+        Assertions.assertThat(w.getMicros(0))
+                .isEqualTo(expectedBaseLatencyMicros);
+    }
+}
```
Comment (jyothsnakonisa) [inline test/unit/org/apache/cassandra/service/TimeoutStrategyTest.java]: Please add new line

File: test/unit/org/apache/cassandra/service/TimeoutStrategyTest.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+import java.util.concurrent.TimeUnit;
+
+public class TimeoutStrategyTest {
+
+    @Test
+    public void testParseLatencyModifierExponential() {
+        long expectedBaseLatencyMicros = TimeUnit.MILLISECONDS.toMicros(30);
+        String spec = "30ms * 2^attempts";
+        TimeoutStrategy.Wait w = org.apache.cassandra.service.TimeoutStrategy.parseWait(spec, TimeoutStrategy.LatencySourceFactory.none());
```
Comment (jyothsnakonisa) [inline test/unit/org/apache/cassandra/service/TimeoutStrategyTest.java]:
```suggestion
        TimeoutStrategy.Wait w = TimeoutStrategy.parseWait(spec, TimeoutStrategy.LatencySourceFactory.none());
```
We don't need fully qualified class name

File: test/unit/org/apache/cassandra/service/TimeoutStrategyTest.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+import org.assertj.core.api.Assertions;
+import org.junit.Test;
+
+import java.util.concurrent.TimeUnit;
+
+public class TimeoutStrategyTest {
+
+    @Test
+    public void testParseLatencyModifierExponential() {
```
Comment (jyothsnakonisa) [inline test/unit/org/apache/cassandra/service/TimeoutStrategyTest.java]: How about checking for Integer.MAX_VALUE & Integer.MIN_VALUE number of attempts for checking overflows?
Also, can you add another case for fractional base like `0.5`

**Review by jyothsnakonisa — COMMENTED** [review]
Looks good overall, just minor comments

**Review by jyothsnakonisa — APPROVED** [review]

File: test/unit/org/apache/cassandra/service/TimeoutStrategyTest.java (hunk 4 above)
Comment (nivykani) [reply test/unit/org/apache/cassandra/service/TimeoutStrategyTest.java]: Great catch with the overflows; Integer.MIN_VALUE would break this logic so I added a max(a, 0). For Integer.MAX_VALUE, the maximum number of retries has to be specified in the input string. Also added a case for fractional bases.

File: src/java/org/apache/cassandra/service/TimeoutStrategy.java [hunk 5]
```diff
@@ -97,7 +97,8 @@ interface LatencyModifierFactory
         default LatencyModifier identity() { return (l, a) -> l; }
         default LatencyModifier multiply(double constant) { return (l, a) -> saturatedCast(l * constant); }
         default LatencyModifier multiplyByAttempts(double multiply) { return (l, a) -> saturatedCast(l * multiply * a); }
-        default LatencyModifier multiplyByAttemptsExp(double base) { return (l, a) -> saturatedCast(l * pow(base, a)); }
+        // Ensure attempts is non-negative before subtracting 1.
+        default LatencyModifier multiplyByAttemptsExp(double base) { return (l, a) -> saturatedCast(l * pow(base, max(0, (max(a, 0) - 1)))); }
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/service/TimeoutStrategy.java:101]: this looks safe

```
$ rg '\^attempts' src
src/java/org/apache/cassandra/config/RetrySpec.java
164:        return RetryStrategy.parse(spec.baseSleepTime.toMilliseconds() + "ms * 2^attempts <= " + spec.maxSleepTime.toMilliseconds() + "ms,retries=" + (spec.maxAttempts.value - 1), LatencySourceFactory.none());

src/java/org/apache/cassandra/service/RetryStrategy.java
72: * <li> {@code 10ms <= p95(rw)*1.8^attempts <= 100ms}

src/java/org/apache/cassandra/service/TimeoutStrategy.java
73: * <li> {@code 10ms <= p95(rw)*1.8^attempts <= 100ms}
```

The only user is repair, so not really a breaking change to paxos.  If anyone was using this in paxos v2, the only impact is that it would have the first retry respect the min value rather than do `min * 2`, so LGTM

File: test/unit/org/apache/cassandra/service/TimeoutStrategyTest.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+ */
+
+package org.apache.cassandra.service;
+
+import org.assertj.core.api.Assertions;
+import org.junit.Test;
+
+import java.util.concurrent.TimeUnit;
+
+public class TimeoutStrategyTest {
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/service/TimeoutStrategyTest.java]: style: `{` are always on a new line

**Review by dcapwell — APPROVED** [review]

**Comment by dcapwell** [comment]
checkstyles fails

File Name	Line Number	Message
/workspace/cassandra/test/unit/org/apache/cassandra/service/TimeoutStrategyTest.java	24	Wrong order for 'java.util.concurrent.TimeUnit' import.

**Comment by dcapwell** [comment]
```
commit 386183fce11707f176b736d92e50cb75ad0680b8
Author: Nivy Kani <nivy@apple.com>
Date:   Thu Jan 8 14:27:36 2026 -0800

    Fix off-by-one bug in exponential backoff for repair retry config

    patch by Nivy Kani; reviewed by David Capwell, Jyothsna Konisa for CASSANDRA-21102
```
